src/Agent.Sdk/SecretMasking/OssSecretMasker.cs

```diff
@@ -81,11 +106,132 @@ public void Dispose()
+    /// rule-based patterns with high entropy if any such detections are found.
+    /// </summary>
+    public void PublishTelemetry(PublishSecretMaskerTelemetryAction publishAction)
+    {
+        if (_telemetry == null)
+        {
+            return;
+        }
+
+        Dictionary<string, string> detectionData = null;
+        int events = 0;
+
+        foreach (var pair in _telemetry.Detections)
+        {
+            detectionData ??= new Dictionary<string, string>(MaxDetectionsPerTelemetryEvent);
+            detectionData.Add(pair.Key, pair.Value);
+
+            if (detectionData.Count >= MaxDetectionsPerTelemetryEvent)
+            {
+                publishAction("SecretMaskerDetections", detectionData);
+                events++;
+                detectionData = null;
+
+                if (events >= MaxTelemetryDetectionEvents)
+                {
+                    break;
+                }
+            }
+        }
+
+        if (events < MaxTelemetryDetectionEvents && detectionData != null)
+        {
+            publishAction("SecretMaskerDetections", detectionData);
+            detectionData = null;
+        }
+
+        var overallData = new Dictionary<string, string>(GetOverallTelemetry());
+        if (_telemetry.Detections.Count > MaxTelemetryDetections)
+        {
+            overallData.Add("DetectionDataIsIncomplete", "true");
+        }
+
+        publishAction("SecretMasker", overallData);
+    }
+
+    private KeyValuePair<string, string>[] GetOverallTelemetry()
+    {
+        double elapsedMaskingTimeInMilliseconds = (double)_secretMasker.ElapsedMaskingTime / TimeSpan.TicksPerMillisecond;
+        return new KeyValuePair<string, string>[] {
+            new("Version", SecretMasker.Version.ToString()),
+            new("CharsScanned", _telemetry.CharsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("StringsScanned", _telemetry.StringsScanned.ToString(CultureInfo.InvariantCulture)),
+            new("TotalDetections", _telemetry.TotalDetections.ToString(CultureInfo.InvariantCulture)),
+            new("UniqueCorrelatingIds", _telemetry.Detections.Count.ToString(CultureInfo.InvariantCulture)),
+            new("ElapsedMaskingTimeInMilliseconds", elapsedMaskingTimeInMilliseconds.ToString(CultureInfo.InvariantCulture)),
+        };
+    }
+
+    private sealed class Telemetry
+    {
+        private readonly ConcurrentDictionary<string, string> _detections;
+        private long _charsScanned;
+        private long _stringsScanned;
+        private long _totalDetections;
+
+        public IReadOnlyDictionary<string, string> Detections => _detections;
+        public long CharsScanned => _charsScanned;
+        public long StringsScanned => _stringsScanned;
+        public long TotalDetections => _totalDetections;
+        public readonly Action<Detection> DetectionAction;
+
+        public Telemetry()
+        {
+            _detections = new ConcurrentDictionary<string, string>();
```

Comment:
>_detections [](http://example.com/codeflow?start=12&length=11)

I don't like this name, it's a SEV3 that I feel SEV2 about, haha. What we've got here is a distinct set of c3ids (which by definition comprise a globally unique value, which therefore implies only a single value moniker exists for each).

`UniqueDetections`? `DeduplicatedDetections`? `C3ids`?

i'm not sure i like any of these names, actually, and so if you want to stick with detections, ok, may just comment the code to note the dictionary comprises the unique set of observed detections as established by their CrossCompanyCorrelatingId.

There's just something about the term 'detections' that implies a stream of raw instances to me.

Reply:
you call similar data `detectionsData` when batching over this dictionary, I note. :)